### PR TITLE
Preserve the bracketed lists in the Parsetree

### PR DESCRIPTION
**Pull request by gpetiot**
This PR keeps the original list form in the parsetree, 2 added constructors are added `Ppat_list` and `Pexp_list`, that mirror the array constructors.
The desugaring is moved to ~`typing/typecore.ml`~`parsing/ast_desugar.ml`, this is a big life improvement for tools like `ocamlformat` as it removes the need to resugar the list to retrieve the original syntax (and will also fix issues related to the placement of comments).

I would like to do the same for empty lists, but for some reason it causes some conflicts on the parser if I add specific rules for `LBRACKET RBRACKET`, maybe the change would be more invasive, I didn't manage to fix that.

### CODE REVIEW

**Review by gasche — COMMENTED** [review]
I understand the motivation and I find it convincing. On first read, not a full review yet, I find the implementation also convincing: it's surprising simple, mostly code moved around and a smaller change than I would have expected. Makes you want to use this approach for more language constructs.

Thanks!

(I'm worried about issues with the constructors not being handled correctly by fragile pattern, but I guess that these would be caught by the testsuite?)

**Comment by alainfrisch** [comment]
I'm in favor of the change.  One point to consider, though, is the impact on ppx that would pattern-match (attribute payloads, for instance) on the current representation for lists, and that would need to be modified to support the new form (or both).

**Comment by nchataing** [comment]
Adding the specific rule for empty list conflicts with the `constr_extra_nonprefix_ident` rule (see https://github.com/ocaml/ocaml/blob/76f390c2001c436f4f9fd2366a646a278340cde9/parsing/parser.mly#L3545), where the empty list is considered as a constructor on its own.

Though after resolving the conflict, and building, we get the following error : 
```
File "/home/void/projects/ocaml/stdlib/listLabels.mli", line 40, characters 22-23:
40 | type 'a t = 'a list = [] | (::) of 'a * 'a list (**)
                           ^
Error: Syntax error
```
In this code (and probably only there), we need to consider `[]` as a constructor...

Maybe there is a workaround ?

**Comment by alainfrisch** [comment]
For the use cases of simplifying ocamlformat, I guess keeping `[]` as a constructor seems relatively harmless.

Otherwise, one possibility would be to use regular constructors for the list type (e.g. Cons and Empty?) instead of `::` and `[]`, and desugar the list literal syntax into `Stdlib.List.Cons / Stdlib.List.Empty` (knowing that `Stdlib.List` re-export constructors from the initial environment).  This would make the grammar more regular (fewer ad hoc cases for valid constructor names).  A minor problem is that it could in theory break user code that use constructors `[]` and `::` in their own types. 

The same could be done for other special constructor names (`()`, `true`, `false`).

**Comment by Octachron** [comment]
>  A minor problem is that it could in theory break user code that use constructors [] and :: in their own types.

This is not a theoretical or minor issue at all and it will definitely break existing libraries (and not only mine).

**Comment by alainfrisch** [comment]
> This is not a theoretical or minor issue at all and it will definitely break existing libraries (and not only mine).

Do you have pointers to code bases that use these constructors in their own types?

**Comment by Gbury** [comment]
> Do you have pointers to code bases that use these constructors in their own types?

Crowbar defines a type that uses the list constructors : https://github.com/stedolan/crowbar/blob/0061f60605b550c967fd5ccdcc95fcab786b55fa/src/crowbar.mli#L6-L10

**Comment by alainfrisch** [comment]
If I understand https://ocaml.org/manual/names.html#constr-name correctly, the fact that `::` and `[]` can be used in custom type declaration is an undocumented behavior.  Also, https://ocaml.org/manual/expr.html#start-section treats these forms as syntactic ones (they are not considered as constructors).

(Not dismissing the actual breakage in the wild, just pointing that this is an undocumented feature.)

**Comment by Octachron** [comment]
Libraries using heterogeneous lists tend to redefine `[]` and `(::)` for instance:

- https://github.com/thierry-martinez/refl/blob/master/runtime/desc.ml#L48
- https://github.com/Octachron/codept/blob/master/lib/zipper_def.ml#L241

Moreover, I don't really see using magical alphabetical identifiers as a progress compared to using explicitly specialized operators like `[]` and `(::)`.

Overall, I think that parsing `[]` as a constructor is the right thing to do.

File: typing/typecore.ml [hunk 1]
```diff
@@ -224,6 +224,49 @@ let rcp node =
   node
 ;;
 
+module Desugar : sig
```
Comment (Octachron) [inline typing/typecore.ml]: Since this module is not really related to typing and is only transforming `Parsetree`s, I think it would make sense to move it to `parsing`.

Comment (gpetiot) [reply typing/typecore.ml]: I've moved this module in parsing/ast_desugar.ml (keeping the same copyright header as parser.mly since the code originates from there)

**Review by gasche — APPROVED** [review]
I reviewed the code and it looks fine.

One difficulty is how the change will affect the ppx ecosystem. One important propertly is that the "desugared" ASTs that we would previously produce are still correct/accepted and will do the right thing -- please stop me if this is wrong! In particular, AST producers/transformers that would produce the desugared form will keep working as before. This suggests that the breakage in the PPX world would be minimal, but still touching the AST means that some changes will be necessary.

**Comment by gasche** [comment]
The testsuite is currently broken:

```
List of failed tests:
    tests/parsetree/'locations_test.ml' with 1 (toplevel) 
```

**Comment by gpetiot** [comment]
> The testsuite is currently broken:
> 
> ```
> List of failed tests:
>     tests/parsetree/'locations_test.ml' with 1 (toplevel) 
> ```

I rebased and re-run the tests locally,  I get no issue when running `make test one DIR=testsuite/parsetree` (3/3 ok), on the CI we got the old parsetree output, is the CI running `make promote-menhir` ?

**Comment by Octachron** [comment]
You need to commit the parser changes: the compiler build process should not depend on the availability of menhir.

**Comment by Drup** [comment]
I'm pretty strongly against this change. This makes some very specialized libraries (pretty much only pretty printers of ocaml code) marginally easier to write, while making several other pieces of code impossible or harder to write. 
1. As @Octachron pointed out, breaking redefined list constructors is very problematic (there is also an example in Eliom, I think in the graphql libraries as well). It might be undocumented, but the fact that it works is rather intuitive, given how the syntax of ocaml behaves in general.
2. It's going to break ppxs *silently* (the problem is not emission of syntax, but pattern matching). The solution will probably involve pushing the recombination @gpetiot wants to avoid doing in ppxlib instead.
3. In general, I think having AST constructs that are sugar of one another is not a good design. We generally avoided it in the past.

All in all, this doesn't seem like the right tradeoff at all.

**Comment by lpw25** [comment]
> breaking redefined list constructors is very problematic

Why would this patch break redefining list constructors? There is no reason for it to do so.

> AST constructs that are sugar of one another is not a good design

Strongly disagree. Having sugar implemented in the parser is not a good design. The AST should reflect what the user wrote. That makes all tooling easier. It makes ppxes easier. It makes good error messages easier.

> It's going to break ppxs silently

I'm not sure how true that is. If people are using ppxlib then my understanding is that they should be matching on lists using the `Ast_pattern.elist` combinator. That combinator simply needs to change its definition when ppx updates its parse tree.

**Comment by Drup** [comment]
> Why would this patch break redefining list constructors? There is no reason for it to do so.

I misunderstood @alainfrisch 's remarks above, it does make the suggested solution invalid though.

> > AST constructs that are sugar of one another is not a good design
> 
> Strongly disagree. Having sugar implemented in the parser is not a good design. The AST should reflect what the user wrote. That makes all tooling easier. It makes ppxes easier. It makes good error messages easier.

The parsetree is not currently designed like that at all. All the syntactic sugar is removed and never represented in the AST (see: type ascriptions, sugar around let bindings, ppx syntax, ....). Designing an AST the other way from a ground up might result in a nice system, but I see no hint it'll improve error message or ppx handling compared to the current situation. The proposed change only gives us an inconsistently designed AST where some sugar is removed, but not all.

> > It's going to break ppxs silently
> 
> I'm not sure how true that is. If people are using ppxlib then my understanding is that they should be matching on lists using the `Ast_pattern.elist` combinator. That combinator simply needs to change its definition when ppx updates its parse tree.

Not everyone uses those combinators (for various reasons, there are still many ppxs that uses ppxlib only as a driver, and pattern match on the AST itself). The metaquot version will also need some magic. As I said, it's probably solvable, but it needs bits of cleverness in ppxlib, which seems of equal complexity to the one in ocamlformat currently, except this time it's user-facing.

**Comment by lpw25** [comment]
>The proposed change only gives us an inconsistently designed AST where some sugar is removed, but not all.

My hope is that we can remove them one at a time until we have a nice AST. Insisting on doing all of them at once will make it much harder to achieve.

**Comment by Drup** [comment]
> > 
> 
> My hope is that we can remove them one at a time until we have a nice AST. Insisting on doing all of them at once will make it much harder to achieve.

If we collectively decide we want to remove them, sure. But that's not what this PR is, and the general direction was never even mentioned. Also, as I said, I'm not even convinced that's generally worth it (just thinking about what the AST is going to look like for `let` makes me anxious).

**Comment by xavierleroy** [comment]
For the record, I agree with @Drup's dissenting opinion.

The proposed AST change would make it possible for a PPX to preprocess `[a; b]` differently from `a :: b :: []`, and it sounds like a bad idea to me.
